lewis/core/logging.py

```diff
@@ -0,0 +1,141 @@
+
+    To provide a more concrete example in terms of Lewis, this is used for the state
+    machine logger in a device. So the logs of the state machine belonging to a certain
+    device appear in the log as originating from ``lewis.DeviceName.StateMachine``, which
+    makes it possible to distinguish between messages from different state machines.
+
+    Furthermore this adds a note to the documentation if the target has any, which appears
+    in the automatically generated documentation of that class or function.
+
+    Example for how to use logging in a class:
+
+    .. sourcecode:: Python
+
+        from lewis.core.logging import has_log
+
+        @has_log
+        class Foo(Base):
+            def __init__(self):
+                super(Foo, self).__init__()
+
+            def bar(self, baz):
+                self.log.debug('Called bar with parameter baz=%s', baz)
+                return baz is not None
+
+    It works similarly for free functions, although the actual logging calls are a bit different:
+
+    .. sourcecode:: Python
+
+        from lewis.core.logging import has_log
+
+        @has_log
+        def foo(bar):
+            foo.log.info('Called with argument bar=%s', bar)
+            return bar
+
+    The name of the logger is ``lewis.foo``, the context could also be modified by calling
+    ``foo._set_logging_context``.
+
+    :param target: Target to decorate with logging functionality.
+
+    .. seealso::
+
+        Two central classes in Lewis inherit from this mixin, :class:`~lewis.adapters.Adapter`
+        and :class:`~lewis.core.devices.DeviceBase`.
+
+        :class:`~lewis.core.statemachine.StateMachine`, :class:`~lewis.core.statemachine.State`
+        and :class:`~lewis.core.statemachine.Transition` also inherit from this, so logging under
+        clearly defined names is automatically available in classes inheriting from those as well.
+    """
+    logger_name = target.__name__
+
+    def get_logger_name(context=None):
+        log_names = [root_logger_name, logger_name]
+
+        if context is not None:
+            log_names.insert(1,
+                             context if isinstance(context,
+                                                   string_types) else context.__class__.__name__)
+
+        return '.'.join(log_names)
+
+    def _set_logging_context(obj, context):
+        """
+        Changes the logger name of this class using the supplied context
+        according to the rules described in the documentation of :func:`has_log`. To
+        clear the context of a class logger, supply ``None`` as the argument.
+
+        :param context: String or object, ``None`` to clear context.
+        """
+        obj.log.name = get_logger_name(context)
+
+    target.log = logging.getLogger(get_logger_name())
+    target._set_logging_context = _set_logging_context
+
+    if target.__doc__ is not None:
+        target.__doc__ += '''
```

Comment:
Interesting, this is not allowed in Python 2, did not realize that. There are two ways around this, the easy one being to change this into a try...except block and configuration in RTD and build with Python 3. This will mean that documentation generated with Python 2 will look different.

There's also the possibility to change this into something slightly more complicated when `target` is a class, I'll play around with that a bit, but let me know if you think the configuration change route would be acceptable.